File: apps/admin/src/modules/devices/chat/chat.gateway.ts
import { ChatSettingsService } from '../../chat-settings/chat-settings.service';
import { ChatTrafficService } from '../../chat-traffic/chat-traffic.service';
import { DEFAULT_CHAT_POLICY } from '../../chat-settings/chat-policy';
import { OnModuleDestroy } from '@nestjs/common';
import { OnGatewayConnection, OnGatewayDisconnect, WebSocketGateway } from '@nestjs/websockets';
import type { RawData } from 'ws';
import WebSocket from 'ws';
import { ChatAuthService } from './chat-auth.service';
import { ChatSessions } from './chat-sessions';
import { CHAT_FRAME_LIMIT, record, send, type ChatIdentity } from './protocol';
import { ChatStunService } from './stun.service';

const POLICY_REFRESH_MS = 5000;
const RATE_WINDOW_MS = 1000;
const MIB = 1024 * 1024;

interface Connection {
  identity?: ChatIdentity;
  authenticating: boolean;
  alive: boolean;
  bytes: number;
  frames: number;
  windowStart: number;
  authTimer: NodeJS.Timeout;
}

@WebSocketGateway({ path: '/device-chat', maxPayload: CHAT_FRAME_LIMIT, perMessageDeflate: false })
export class ChatGateway implements OnGatewayConnection, OnGatewayDisconnect, OnModuleDestroy {
  private readonly connections = new Map<WebSocket, Connection>();
  private readonly sessions: ChatSessions;
  private readonly heartbeat = setInterval(() => this.tick(), 25_000);

  private refreshingPolicy = false;
  private policy = { ...DEFAULT_CHAT_POLICY };
  private readonly policyTimer = setInterval(() => { void this.refreshPolicy(); }, POLICY_REFRESH_MS);

  constructor(private readonly auth: ChatAuthService, private readonly stun: ChatStunService,
    private readonly settings: ChatSettingsService, traffic: ChatTrafficService) {
    this.sessions = new ChatSessions((bytes) => traffic.record(bytes));
    this.policyTimer.unref();
    this.heartbeat.unref();
  }

  handleConnection(client: WebSocket) {
    const authTimer = setTimeout(() => client.close(4001, 'Authentication timed out'), 10_000);
    this.connections.set(client, {
      authenticating: false, alive: true, bytes: 0, frames: 0, windowStart: Date.now(), authTimer,
    });
    client.on('pong', () => { const state = this.connections.get(client); if (state) state.alive = true; });
    client.on('message', (raw: RawData, binary: boolean) => {
      void this.receive(client, raw, binary).catch(() => {
        this.sessions.disconnect(client, true);
        client.close(4001, 'Chat connection rejected');
      });
    });
    client.once('close', () => this.handleDisconnect(client));
  }

  private async receive(client: WebSocket, raw: RawData, binary: boolean) {
    const state = this.connections.get(client);
    const size = Array.isArray(raw) ? raw.reduce((total, part) => total + part.length, 0) : raw.byteLength;
    if (!state || binary || size > CHAT_FRAME_LIMIT) throw new Error('Invalid frame');
    this.checkRate(state, size);
    const message = record(JSON.parse(raw.toString()) as unknown);
    if (state.identity) {
      if (state.identity.expiresAt <= Date.now()) throw new Error('Expired');
      this.sessions.route(client, message);
      return;
    }
    if (state.authenticating) throw new Error('Authentication pending');
    state.authenticating = true;
    const identity = await this.auth.authenticate(message);
    const policy = await this.settings.read();
    this.policy = policy;
    if (client.readyState !== WebSocket.OPEN || !this.connections.has(client)) return;
    state.identity = identity;
    clearTimeout(state.authTimer);
    state.authTimer = setTimeout(() => {
      this.sessions.disconnect(client, true);
      client.close(4001, 'Session expired');
    }, identity.expiresAt - Date.now());
    send(client, { type: 'chat-policy', policy });
    this.sessions.join(client, identity, message, this.stun.iceServers());
  }

  private async refreshPolicy() {
    if (this.refreshingPolicy || !this.connections.size) return;
    this.refreshingPolicy = true;
    try {
      const policy = await this.settings.read();
      this.policy = policy;
      for (const [client, state] of this.connections) {
        if (state.identity && state.identity.expiresAt > Date.now()) send(client, { type: 'chat-policy', policy });
      }
    } catch { /* Keep the last confirmed policy during a temporary database outage. */ }
    finally { this.refreshingPolicy = false; }
  }

  private checkRate(state: Connection, size: number) {
    if (Date.now() - state.windowStart >= RATE_WINDOW_MS) {
      state.bytes = 0;
      state.frames = 0;
      state.windowStart = Date.now();
    }
    state.bytes += size;
    state.frames += 1;
    const { relayMaxMbPerSecond, relayMaxFramesPerSecond } = this.policy;
    if ((relayMaxMbPerSecond !== -1 && state.bytes / MIB > relayMaxMbPerSecond)
      || (relayMaxFramesPerSecond !== -1 && state.frames > relayMaxFramesPerSecond)) {
      throw new Error('Rate exceeded');
    }
  }

  private tick() {
    this.sessions.prune();
    for (const [client, state] of this.connections) {
      if (!state.alive) { client.terminate(); continue; }
      state.alive = false;
      if (client.readyState === WebSocket.OPEN) client.ping();
    }
  }

  handleDisconnect(client: WebSocket) {
    const state = this.connections.get(client);
    if (state) clearTimeout(state.authTimer);
    this.connections.delete(client);
    this.sessions.disconnect(client);
  }

  onModuleDestroy() {
    clearInterval(this.heartbeat);
    clearInterval(this.policyTimer);
    for (const client of this.connections.keys()) {
      this.handleDisconnect(client);
      client.close(1001, 'Server shutting down');
    }
  }
}

File: apps/admin-go/internal/devices/chat_gateway.go
package devices

import (
	"errors"
	"log/slog"
	"sync"
	"time"

	"github.com/codex-switch/admin-go/internal/chattraffic"
	"github.com/codex-switch/admin-go/internal/content"
	"github.com/codex-switch/admin-go/internal/platform"
	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
)

type chatConnection struct {
	identity      *chatIdentity
	bytes, frames int
	windowStart   time.Time
	timer         *time.Timer
}
type ChatGateway struct {
	meter       *chattraffic.Meter
	service     *Service
	sessions    *chatSessions
	traffic     *trafficCounter
	mu          sync.Mutex
	policyMu    sync.Mutex
	connections map[*peer]*chatConnection
	policy      platform.JSON
	ice         []platform.JSON
	done        chan struct{}
	stopped     chan struct{}
	stun        *stunServer
}

func newChatGateway(service *Service) (*ChatGateway, error) {
	ice, err := iceServers(service.deps.Config)
	if err != nil {
		return nil, err
	}
	stun, err := startSTUN(service.deps.Config)
	if err != nil {
		return nil, err
	}
	traffic := newTrafficCounter(service.deps.DB)
	gateway := &ChatGateway{service: service, sessions: newChatSessions(traffic.record), traffic: traffic,
		meter:       chattraffic.NewMeter(service.deps.DB, service.deps.Redis),
		connections: map[*peer]*chatConnection{}, policy: platform.JSON{"relayMaxMbPerSecond": float64(-1),
			"relayMaxFramesPerSecond": float64(
				-1,
			)}, ice: ice, done: make(chan struct{}), stopped: make(chan struct{}), stun: stun}
	service.deps.FlushTraffic = traffic.flush
	gateway.sessions.deliver = gateway.deliverRelay
	gateway.sessions.hot.deliver = gateway.deliverRelay
	service.deps.ChatPolicyChanged = gateway.refreshPolicy
	go gateway.maintain()
	return gateway, nil
}

func (g *ChatGateway) serve(c *gin.Context) {
	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		return
	}
	conn.SetReadLimit(chatFrameLimit)
	client := newPeerWithDiagnostics(conn, newChatDiagnostics())
	state := &chatConnection{windowStart: time.Now()}
	state.timer = time.AfterFunc(authTimeout, func() { client.close(4001, "Authentication timed out") })
	g.mu.Lock()
	g.connections[client] = state
	g.mu.Unlock()
	defer g.disconnect(client)
	for {
		kind, data, err := conn.ReadMessage()
		if err != nil {
			code := websocket.CloseAbnormalClosure
			var closed *websocket.CloseError
			if errors.As(err, &closed) {
				code = closed.Code
			}
			client.diagnostics.close(code, "socket read ended")
			return
		}
		client.diagnostics.received(len(data))
		message, err := readChatFrame(client, kind, data)
		if err == nil {
			err = g.receive(client, state, message, len(data))
		}
		if err != nil {
			client.diagnostics.log("chat rejected", "reason", chatRejectionReason(err))
			g.reject(client)
			return
		}
	}
}

func (g *ChatGateway) receive(client *peer, state *chatConnection, message platform.JSON, size int) error {
	if err := g.checkRate(state, size); err != nil {
		return err
	}
	if state.identity != nil {
		if !state.identity.expires.After(time.Now()) {
			return errors.New("expired token")
		}
		return g.sessions.route(client, message)
	}
	identity, err := g.authenticate(message)
	if err != nil {
		return err
	}
	client.diagnostics.authenticated(identity, message)
	client.binaryRelay.Store(message["binaryRelay"] == true)
	// Serialize handshake snapshots with saved-policy broadcasts so stale reads cannot follow a new policy.
	g.policyMu.Lock()
	defer g.policyMu.Unlock()
	policy, err := content.ReadChatPolicy(g.service.deps.DB)
	if err != nil {
		return err
	}
	g.mu.Lock()
	if client.closed.Load() {
		g.mu.Unlock()
		return errors.New("closed connection")
	}
	g.policy = policy
	state.identity = &identity
	state.timer.Stop()
	state.timer = time.AfterFunc(time.Until(identity.expires), func() {
		g.sessions.disconnect(client, true)
		client.close(4001, "Session expired")
	})
	g.mu.Unlock()
	client.send(platform.JSON{"type": "chat-policy", "policy": policy,
		"binaryRelay": client.binaryRelay.Load()}, nil)
	g.sessions.setLimit(policy["chatSessionLimit"].(float64))
	return g.sessions.join(client, identity, message, g.ice)
}

func (g *ChatGateway) authenticate(message platform.JSON) (chatIdentity, error) {
	identity := chatIdentity{}
	token, ok := message["accessToken"].(string)
	if !ok || len(token) > 8192 || message["type"] != "authenticate" {
		return identity, errors.New("invalid authentication")
	}
	role, _ := message["role"].(string)
	if role != "desktop" && role != "mobile" {
		return identity, errors.New("invalid role")
	}
	owner, expires, err := socketIdentity(g.service.deps, token)
	if err != nil || !expires.After(time.Now()) {
		return identity, errors.New("expired token")
	}
	id, err := identifier(message["deviceId"])
	if err != nil {
		return identity, err
	}
	if _, err := g.service.owned(owner, id); err != nil {
		return identity, err
	}
	return chatIdentity{owner, id, role, expires}, nil
}

func (g *ChatGateway) checkRate(state *chatConnection, size int) error {
	if time.Since(state.windowStart) >= time.Second {
		state.bytes, state.frames, state.windowStart = 0, 0, time.Now()
	}
	state.bytes += size
	state.frames++
	g.mu.Lock()
	megabytes, frames := policyNumber(
		g.policy["relayMaxMbPerSecond"],
	), policyNumber(
		g.policy["relayMaxFramesPerSecond"],
	)
	g.mu.Unlock()
	if (megabytes != -1 && float64(state.bytes)/(1024*1024) > megabytes) ||
		(frames != -1 && float64(state.frames) > frames) {
		return errors.New("rate exceeded")
	}
	return nil
}

func policyNumber(value interface{}) float64 {
	switch value := value.(type) {
	case float64:
		return value
	case int:
		return float64(value)
	default:
		return -1
	}
}

func (g *ChatGateway) reject(client *peer) {
	g.sessions.disconnect(client, true)
	client.close(4001, "Chat connection rejected")
}

func (g *ChatGateway) disconnect(client *peer) {
	g.mu.Lock()
	if state := g.connections[client]; state != nil {
		state.timer.Stop()
	}
	delete(g.connections, client)
	g.mu.Unlock()
	client.terminate()
	g.sessions.disconnect(client, false)
}

func (g *ChatGateway) maintain() {
	defer close(g.stopped)
	timer := time.NewTicker(5 * time.Second)
	defer timer.Stop()
	for {
		select {
		case <-g.done:
			return
		case <-timer.C:
			g.sessions.prune()
			if err := g.meter.Flush(); err != nil {
				slog.Warn("relay accounting settlement will be retried", "error", err)
			}
			g.publishTraffic()
			g.refreshPolicy()
			if err := g.traffic.flush(); err != nil {
				slog.Warn("chat traffic flush will be retried", "error", err)
			}
		}
	}
}

func (g *ChatGateway) refreshPolicy() {
	g.policyMu.Lock()
	defer g.policyMu.Unlock()
	g.mu.Lock()
	active := len(g.connections) > 0
	g.mu.Unlock()
	if !active {
		return
	}
	policy, err := content.ReadChatPolicy(g.service.deps.DB)
	if err != nil {
		slog.Warn("chat policy refresh will be retried", "error", err)
		return
	} // Keep the last confirmed policy during a database outage.
	g.sessions.setLimit(policy["chatSessionLimit"].(float64))
	g.mu.Lock()
	defer g.mu.Unlock()
	g.policy = policy
	for client, state := range g.connections {
		if state.identity != nil && state.identity.expires.After(time.Now()) {
			client.send(platform.JSON{"type": "chat-policy", "policy": policy}, nil)
		}
	}
}

type Runtime struct {
	control *ControlGateway
	chat    *ChatGateway
}

func (runtime *Runtime) Close() error {
	runtime.control.mu.Lock()
	for client := range runtime.control.sessions {
		client.close(1001, "Server shutting down")
	}
	runtime.control.mu.Unlock()
	g := runtime.chat
	close(g.done)
	<-g.stopped
	g.mu.Lock()
	for client, state := range g.connections {
		state.timer.Stop()
		client.close(1001, "Server shutting down")
	}
	g.mu.Unlock()
	if g.stun != nil {
		g.stun.close()
	}
	return errors.Join(g.meter.Close(), g.traffic.flush())
}
